src/actions/zipped-profiles.js

```diff
@@ -0,0 +1,107 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+import { getZipFileTable, getZipFileState } from '../reducers/zipped-profiles';
+import { unserializeProfileOfArbitraryFormat } from '../profile-logic/process-profile';
+import type { Action, ThunkAction } from '../types/store';
+import type { IndexIntoZipFileTable } from '../profile-logic/zip-files';
+
+export function changeSelectedZipFile(
+  selectedZipFileIndex: IndexIntoZipFileTable
+): Action {
+  return {
+    type: 'CHANGE_SELECTED_ZIP_FILE',
+    selectedZipFileIndex,
+  };
+}
+
+export function changeExpandedZipFile(
+  expandedZipFileIndexes: Array<IndexIntoZipFileTable | null>
+): Action {
+  return {
+    type: 'CHANGE_EXPANDED_ZIP_FILES',
+    expandedZipFileIndexes,
+  };
+}
+
+/**
+ * This ThunkAction has a bit of complexity to it, due to the asynchronous nature of
+ * reading in profiles from a zip file. The UrlState represents the current desired state
+ * of what file to view in a zip file, and the ZipFileState represents the actual steps
+ * being performed to reach the desired UrlState.
+ *
+ * This ThunkAction needs to properly handle when the UrlState changes faster than it
+ * can change the ZipFileState, and not have any race conditions.
+ */
+export function viewProfileFromZip(
+  zipFileIndex: IndexIntoZipFileTable
+): ThunkAction<Promise<void>> {
+  return async (dispatch, getState) => {
+    const zipFileTable = getZipFileTable(getState());
+    const pathInZipFile = zipFileTable.path[zipFileIndex];
+    const file = zipFileTable.file[zipFileIndex];
+    if (!file) {
+      throw new Error(
+        'Attempted to load a zip file that did not exist or was a directory.'
+      );
+    }
+
+    dispatch({ type: 'PROCESS_PROFILE_FROM_ZIP_FILE', pathInZipFile });
+
+    try {
+      // Attempt to unserialize the profile.
+      const profile = await unserializeProfileOfArbitraryFormat(
```

Comment:
nit: `unserializeProfileOfArbitraryFormat` isn't asynchronous; you might want to remove this `await`. (although it does no harm)